Bottom: 5b5114d55bfe47e2faea9759ee0711168cd4a684
Top:    145eefc0f8e917487b91e8c935627783b0676695
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-05-21 16:00:31 +0800

Refresh of virtio-net-fix-leaking-page-0

---

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 75a0368..9d2da55 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -767,7 +767,9 @@ static struct sk_buff *receive_mergeable(struct net_device *dev,
 			 * adjusted
 			 */
 			len = xdp.data_end - xdp.data + vi->hdr_len;
-			break;
+			truesize = PAGE_SIZE;
+			rcu_read_unlock();
+			goto create_skb;
 		case XDP_TX:
 			xdpf = convert_to_xdp_frame(&xdp);
 			if (unlikely(!xdpf))
@@ -805,6 +807,7 @@ static struct sk_buff *receive_mergeable(struct net_device *dev,
 		goto err_skb;
 	}
 
+create_skb:
 	head_skb = page_to_skb(vi, rq, page, offset, len, truesize);
 	curr_skb = head_skb;
